presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CrossJoinWithArrayContainsToInnerJoin.java

```diff
@@ -0,0 +1,181 @@
+import static com.facebook.presto.expressions.LogicalRowExpressions.and;
+import static com.facebook.presto.expressions.LogicalRowExpressions.extractConjuncts;
+import static com.facebook.presto.matching.Capture.newCapture;
+import static com.facebook.presto.sql.planner.plan.Patterns.filter;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.facebook.presto.sql.relational.Expressions.call;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Inner join with contains function inside join clause will be run as cross join with filter.
+ * When the join condition has pattern of contains(array, element), we can rewrite it to a inner join. For example:
+ * <pre>
+ * - Filter contains(l_array_key, r_key)
+ *      - Cross join
+ *          - scan l
+ *          - scan r
+ * </pre>
+ * into:
+ * <pre>
+ *     - Join
+ *          field = r_key
+ *          - Unnest
+ *              field <- unnest distinct_array
+ *              - project
+ *                  distinct_array := array_distinct(l_array_key)
+ *                  - scan l
+ *                      l_array_key
+ *              - scan r
+ *                  r_key
+ * </pre>
+ */
+public class CrossJoinWithArrayContainsToInnerJoin
+        implements Rule<FilterNode>
+{
+    private static final List<Type> SUPPORTED_JOIN_KEY_TYPE = ImmutableList.of(BIGINT, INTEGER, VARCHAR, DATE);
+    private static final Capture<JoinNode> CHILD = newCapture();
+    private static final Pattern<FilterNode> PATTERN = filter()
+            .with(source().matching(join().matching(x -> x.getType().equals(JoinNode.Type.INNER) && x.getCriteria().isEmpty()).capturedAs(CHILD)));
+
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public CrossJoinWithArrayContainsToInnerJoin(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+    }
+
+    public static RowExpression getCandidateArrayContainsExpression(RowExpression filterPredicate, List<VariableReferenceExpression> leftInput, List<VariableReferenceExpression> rightInput)
+    {
+        List<RowExpression> andConjuncts = extractConjuncts(filterPredicate);
+        for (RowExpression conjunct : andConjuncts) {
+            if (isValidArrayContainsFilter(conjunct, leftInput, rightInput)) {
+                return conjunct;
+            }
+        }
+        return null;
+    }
+
+    private static boolean isValidArrayContainsFilter(RowExpression filterExpression, List<VariableReferenceExpression> left, List<VariableReferenceExpression> right)
+    {
+        if (filterExpression instanceof CallExpression) {
+            CallExpression callExpression = (CallExpression) filterExpression;
+            if (callExpression.getDisplayName().equals("contains")) {
```

Comment:
There should be a better way to do this (like using function handle or something)?

Reply:
Also maybe best to check that array element type and the second argument type are same (assuming any implicit casting needed is already added)